### PR TITLE
Add safe_normalize to nn.functional and use it for mesh normals

### PR DESCRIPTION
**Pull request by peterdsharpe**
# PhysicsNeMo Pull Request

## Description

The fixed epsilon in `torch.nn.functional.normalize` can produce NaNs for zero `float16` vectors and shorten small nonzero vectors. Computing the norm without rescaling can also overflow for large finite inputs and return a zero vector.

Adds `physicsnemo.nn.functional.safe_normalize(vectors, dim)`, which rescales each vector by its largest absolute component before computing its L2 norm. It preserves exactly-zero vectors and the input dtype, including under autocast.

- Uses the shared functional for cell, point, partition-cluster, and transformed cached mesh normals.
- Replaces boolean-mask assignment in the partition centroid fallback with `torch.where` to avoid a host synchronization.
- Adds API documentation and records the new functional and mesh fixes under 2.3.0 in the changelog.

Rescaling adds a pass over the vectors; mesh normals remain cached. Non-finite inputs propagate NaNs, and derivatives near zero can still exceed the dtype's range.

## Verification

Validated commit `df82868d7086816384dd9395f599f8af1911241d`:

- **295 focused tests passed, 1 skipped**, with CPU/CUDA coverage for autocast dtype preservation, noncontiguous inputs and axis selection, mesh normals, partitioning, transformations, shape gradients, and synchronization-free linear algebra. Tests also cover zero vectors and extreme representable magnitudes across floating-point dtypes.
- The `nn.functional.normalization` doctest passed.
- All changed-file pre-commit hooks passed, including import-layer checks.

The full mesh suite and performance benchmarks were not rerun.

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [x] Model implementation standards are not applicable; no model code is changed.

## Dependencies

None.

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score. This score reflects the AI's assessment of merge readiness and is **not** a qualitative judgment of the work or an indication that the PR will be accepted or rejected.

AI-generated feedback should be reviewed critically for usefulness. You are not required to respond to every AI comment, but they are intended to help both authors and reviewers. Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -667,6 +667,21 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   broke the `test_get_checkpoint_dir` CI test on Windows. The function now
   always joins with `/`, working uniformly for local paths and `fsspec`
   URIs (`msc://`, etc.) across operating systems.
+- Mesh normals no longer inherit the hardcoded `eps=1e-12` clamp of
```
Comment (laserkelvin) [inline CHANGELOG.md]: Need to rebase this on the new release?

File: physicsnemo/mesh/utilities/_tolerances.py [hunk 2]
```diff
@@ -76,3 +79,83 @@ def safe_eps(dtype: torch.dtype) -> float:
     """
     info = torch.finfo(dtype)
     return min(info.tiny**0.25, info.eps)
+
+
+def safe_normalize(
```
Comment (laserkelvin) [inline physicsnemo/mesh/utilities/_tolerances.py]: I'm not sure that this *needs* to be under mesh utilities - if anything, maybe living under functionals could be a good home for it

Some places (including `mesh`) that could potentially benefit from `safe_normalize`:

Image

File: physicsnemo/mesh/utilities/_tolerances.py [hunk 3]
```diff
@@ -76,3 +79,83 @@ def safe_eps(dtype: torch.dtype) -> float:
...
+    ### ``amax`` rejects a zero-size reduction dim, and there is nothing to
+    ### normalize in that case anyway.
+    if vectors.shape[dim] == 0:
+        return vectors
+
+    ### Rescaling adds a pass over ``vectors`` relative to a plain clamp and
+    ### divide, making this memory-bound kernel roughly 2-4x slower on its own.
+    ### That is a deliberate correctness-over-speed trade, not an oversight:
+    ### removing the rescale reintroduces the overflow and small-norm errors
+    ### described above, which ``TestSafeNormalize`` pins down. Mesh normals
```
Comment (laserkelvin) [inline physicsnemo/mesh/utilities/_tolerances.py]: I feel like this comment is unnecessarily detailed, I think some intention documented is good but it's too specific to a particular usecase

**Review by laserkelvin — COMMENTED** [review]
I'm generally fine with the PR itself, but I would suggest you consider putting `safe_normalize` and accompanying things in a broader namespace instead

If you want to leave that out of scope for this PR, I'll approve (but consider it though)

File: physicsnemo/mesh/utilities/_tolerances.py (hunk 2 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/utilities/_tolerances.py]: This is a good point, let me take a look at putting it in a higher-level utils folder

File: physicsnemo/mesh/utilities/_tolerances.py (hunk 3 above)
Comment (peterdsharpe) [reply physicsnemo/mesh/utilities/_tolerances.py]: Good call, my bad - should have caught this

**Review by laserkelvin — APPROVED** [review]
LGTM!

File: physicsnemo/nn/functional/normalization.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Vector normalization across floating-point dtypes and scales."""
+
+import torch
+from jaxtyping import Float
+
+
+def safe_normalize(
```
Comment (loliverhennigh) [inline physicsnemo/nn/functional/normalization.py]: Ideally we use the FunctionSpec here.

**Review by loliverhennigh — APPROVED** [review]
Left a comment but good otherwise

File: physicsnemo/nn/functional/normalization.py (hunk 4 above)
Comment (peterdsharpe) [reply physicsnemo/nn/functional/normalization.py]: Will do! Pushing...

**Comment by peterdsharpe** [comment]
/ok to test https://github.com/NVIDIA/physicsnemo/pull/1883/commits/04873e9b5f4dd5710dadde34ab1e244fa3ad1e6a
